// ngforage/esm5/lib/instance-factory/instance-factory.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/instance-factory/instance-factory.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import 'localforage';
import { NgForageConfig } from '../config/ng-forage-config.service';
import { localForage as lf } from '../imports/localforage';
import * as i0 from "@angular/core";
import * as i1 from "../config/ng-forage-config.service";
/**
 * \@internal
 * @record
 */
function InstanceMap() { }
/**
 * \@internal
 * @type {?}
 */
var stores = {};
/**
 * \@internal
 * @param {?=} driver
 * @return {?}
 */
function getDriverString(driver) {
    if (!driver) {
        return '';
    }
    else if (Array.isArray(driver)) {
        return driver.slice().sort().join(',');
    }
    else {
        return (/** @type {?} */ (driver));
    }
}
/**
 * \@internal
 * @param {?} cfg
 * @return {?}
 */
function getHash(cfg) {
    return [
        getDriverString(cfg.driver),
        cfg.name,
        cfg.size,
        cfg.storeName,
        cfg.version,
        cfg.description,
        cfg.cacheTime
    ].join('|');
}
/**
 * \@internal
 * @type {?}
 */
var conf$ = Symbol('Config');
/**
 * Creates localForage instances
 */
var InstanceFactory = /** @class */ (function () {
    /** @internal */
    function InstanceFactory(conf) {
        this[conf$] = conf;
    }
    /**
     * @param {?} cfg
     * @return {?}
     */
    InstanceFactory.prototype.getInstance = /**
     * @param {?} cfg
     * @return {?}
     */
    function (cfg) {
        cfg = Object.assign({}, this[conf$].config, cfg || {});
        /** @type {?} */
        var hash = getHash(cfg);
        if (!stores[hash]) {
            stores[hash] = lf.createInstance(cfg);
        }
        return stores[hash];
    };
    InstanceFactory.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    InstanceFactory.ctorParameters = function () { return [
        { type: NgForageConfig }
    ]; };
    /** @nocollapse */ InstanceFactory.ɵprov = i0.ɵɵdefineInjectable({ factory: function InstanceFactory_Factory() { return new InstanceFactory(i0.ɵɵinject(i1.NgForageConfig)); }, token: InstanceFactory, providedIn: "root" });
    return InstanceFactory;
}());
export { InstanceFactory };
