// neutrinos-seed-services/esm2015/src/app/neutrinos-seed-services/snackbar/snack/snack.component.js
import { Component, Input } from '@angular/core';
export class SnackComponent {
}
SnackComponent.decorators = [
    { type: Component, args: [{
                selector: 'n-snack',
                template: `
     <div class="snack" [ngStyle]="{background: background ? background : '#323232' , color: color ? color : '#fff'}" [ngClass]=" ['snackbar-layout', 'customClass' ]">
          <ng-content></ng-content>
       </div> 
  `,
                styles: [".snack-action{color:#2196f3;cursor:pointer;font-weight:700}.container{display:flex;flex-direction:row;flex-wrap:wrap}.child{flex:1 0;margin-right:.5em;width:60%}.snackbars.bottom-center{align-items:middle}.snackbars{display:flex;display:inline;font-family:Roboto;letter-spacing:1px;max-width:100%;position:fixed;word-wrap:break-word;z-index:99999}.snackbars.bottom-center{align-items:center;bottom:1px;left:50%;transform:translate(-50%)}.snack{border-radius:.3em;max-height:50vh;overflow:auto;padding:1em}"]
            },] }
];
SnackComponent.propDecorators = {
    background: [{ type: Input }],
    color: [{ type: Input }],
    customClass: [{ type: Input }]
};
