// neutrinos-module/esm2015/src/app/neutrinos-module/neutrinos-file/directives/n-shake.directive.js
import { __decorate } from "tslib";
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { HostListener } from '@angular/core';
import { NFileIOService } from '../services/n-fileIO.service';
let NShakeDirective = class NShakeDirective {
    constructor(fsv) {
        this.fsv = fsv;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    methodToHandleMouseClickAction() {
        this.getShake()
            .then(sucess => this.onsuccess.emit(sucess))
            .catch(error => this.onerror.emit(error));
    }
    getShake() {
        return new Promise((resolve, reject) => {
            this.fsv.getShake(this.shakeOptions).then(res => {
                return resolve(res);
            }).catch(err => reject(err));
        });
    }
};
NShakeDirective.ctorParameters = () => [
    { type: NFileIOService }
];
__decorate([
    Input()
], NShakeDirective.prototype, "shakeOptions", void 0);
__decorate([
    Output()
], NShakeDirective.prototype, "onsuccess", void 0);
__decorate([
    Output()
], NShakeDirective.prototype, "onerror", void 0);
__decorate([
    HostListener('click')
], NShakeDirective.prototype, "methodToHandleMouseClickAction", null);
NShakeDirective = __decorate([
    Directive({
        selector: '[n-shake]'
    })
], NShakeDirective);
export { NShakeDirective };
