// neutrinos-module/esm2015/src/app/neutrinos-module/neutrinos-file/directives/n-tts.directive.js
import { __decorate } from "tslib";
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { HostListener } from '@angular/core';
import { NFileIOService } from '../services/n-fileIO.service';
let NTtsDirective = class NTtsDirective {
    constructor(fsv) {
        this.fsv = fsv;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    methodToHandleMouseClickAction() {
        this.getTts()
            .then(sucess => this.onsuccess.emit(sucess))
            .catch(error => this.onerror.emit(error));
    }
    getTts() {
        return new Promise((resolve, reject) => {
            this.fsv.getTts(this.ttsOptions).then(res => {
                return resolve(res);
            }).catch(err => reject(err));
        });
    }
};
NTtsDirective.ctorParameters = () => [
    { type: NFileIOService }
];
__decorate([
    Input()
], NTtsDirective.prototype, "ttsOptions", void 0);
__decorate([
    Output()
], NTtsDirective.prototype, "onsuccess", void 0);
__decorate([
    Output()
], NTtsDirective.prototype, "onerror", void 0);
__decorate([
    HostListener('click')
], NTtsDirective.prototype, "methodToHandleMouseClickAction", null);
NTtsDirective = __decorate([
    Directive({
        selector: '[n-tts]'
    })
], NTtsDirective);
export { NTtsDirective };
