// neutrinos-oauth-client/esm2015/lib/cookie-interceptor.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/cookie-interceptor.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { throwError } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { UtilService } from './services/util.service';
import { NeutrinosOAuthClientService } from './neutrinos-oauth-client.service';
import * as i0 from "@angular/core";
import * as i1 from "./services/util.service";
import * as i2 from "./neutrinos-oauth-client.service";
export class CookieInterceptorService {
    /**
     * @param {?} utilService
     * @param {?} neutrinosOAuthClientService
     */
    constructor(utilService, neutrinosOAuthClientService) {
        this.utilService = utilService;
        this.neutrinosOAuthClientService = neutrinosOAuthClientService;
    }
    /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    intercept(request, next) {
        /* withCredentials adds cookies to the request for cross site (Only ssd urls).*/
        /** @type {?} */
        const baseUrl = this.utilService.getSSDBasePath();
        /** @type {?} */
        const isSSD = (baseUrl !== '' && request.url.includes(baseUrl));
        if (isSSD && this.isCrossDomain(baseUrl)) {
            request = request.clone({
                withCredentials: true
            });
        }
        return next.handle(request).pipe(map((/**
         * @param {?} event
         * @return {?}
         */
        event => {
            return event;
        })), catchError((/**
         * @param {?} errorResponse
         * @return {?}
         */
        (errorResponse) => {
            /*Intercept all requests and remove userinfo and cookies when status is 401 with code SESSION_EXPIRED, which means
            * cookies have expired or access has been revoked by the OAuth providers hence IDS rejected.  */
            if (errorResponse && errorResponse.hasOwnProperty('status') && errorResponse.status === 401 &&
                errorResponse.error.hasOwnProperty('code')) {
                this.neutrinosOAuthClientService.removeUserInfo(errorResponse.error);
            }
            return throwError(errorResponse);
        })));
    }
    /**
     * Returns true is the ssd url is running on a different domain than where UI is hosted.
     * For mobile Returns false always since UI is already bundled with cordova
     * If the user is not using IDS and is cross domain isIDSEnabled can be set to false to skip adding withCredentials
     * @private
     * @param {?} ssdUrl - string
     * @return {?}
     */
    isCrossDomain(ssdUrl) {
        try {
            ssdUrl = new URL(ssdUrl);
            return (this.utilService.getPlatformType() === 'browser' && (ssdUrl.host !== location.host)
                && this.utilService.getEnvValue('isIDSEnabled') === 'true');
        }
        catch (e) {
            return false; // ssdUrl is not a valid URL hence return false;
        }
    }
}
CookieInterceptorService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
CookieInterceptorService.ctorParameters = () => [
    { type: UtilService },
    { type: NeutrinosOAuthClientService }
];
/** @nocollapse */ CookieInterceptorService.ɵprov = i0.ɵɵdefineInjectable({ factory: function CookieInterceptorService_Factory() { return new CookieInterceptorService(i0.ɵɵinject(i1.UtilService), i0.ɵɵinject(i2.NeutrinosOAuthClientService)); }, token: CookieInterceptorService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    CookieInterceptorService.prototype.utilService;
    /**
     * @type {?}
     * @private
     */
    CookieInterceptorService.prototype.neutrinosOAuthClientService;
}
