// neutrinos-oauth-client/esm2015/lib/neutrinos-oauth.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/neutrinos-oauth.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { NeutrinosOAuthClientService } from './neutrinos-oauth-client.service';
import { HttpClient } from '@angular/common/http';
export class NeutrinosOauthComponent {
    /**
     * @param {?} neutrinosOauthService
     * @param {?} http
     */
    constructor(neutrinosOauthService, http) {
        this.neutrinosOauthService = neutrinosOauthService;
        this.http = http;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.neutrinosOauthService.authState().subscribe((/**
         * @param {?} state
         * @return {?}
         */
        state => {
            console.log(state);
        }));
    }
    /**
     * @return {?}
     */
    login() {
        /** @type {?} */
        const a = this.neutrinosOauthService.login('/i-am-loggedin').then((/**
         * @param {?} tokens
         * @return {?}
         */
        tokens => {
            console.log('got tokens in component', tokens);
        })).catch((/**
         * @param {?} e
         * @return {?}
         */
        e => {
            console.error(e);
        }));
        // console.log(a);
    }
    /**
     * @return {?}
     */
    token() {
        this.neutrinosOauthService.getUserInfo().then((/**
         * @param {?} t
         * @return {?}
         */
        t => {
            console.log('t', t);
        })).catch(console.log);
    }
    /**
     * @return {?}
     */
    logout() {
        this.neutrinosOauthService.logout();
    }
    /**
     * @return {?}
     */
    userinfo() {
        alert(JSON.stringify(this.neutrinosOauthService.userInfo));
        console.log(this.neutrinosOauthService.userInfo);
    }
    /**
     * @return {?}
     */
    httpTest() {
        this.http.get('http://localhost:8081/api/ping').subscribe((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            console.log(res);
        }), (/**
         * @param {?} e
         * @return {?}
         */
        e => {
            console.error(e);
        }));
    }
}
NeutrinosOauthComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-neutrinos-oauth',
                template: `
   NOTE: use this component only for testing. Not intended for production

      <button (click)="login()" >  Login </button>
      <button (click)="token()" > Get tokenset</button>
      <button (click)="logout()" > Logout </button>
      <button (click)="userinfo()" > In memory user info </button>
      <button (click)="httpTest()" > Http Interceptor test </button>
      `
            }] }
];
/** @nocollapse */
NeutrinosOauthComponent.ctorParameters = () => [
    { type: NeutrinosOAuthClientService },
    { type: HttpClient }
];
if (false) {
    /** @type {?} */
    NeutrinosOauthComponent.prototype.neutrinosOauthService;
    /**
     * @type {?}
     * @private
     */
    NeutrinosOauthComponent.prototype.http;
}
