// neutrinos-module/esm5/src/app/neutrinos-module/neutrinos-file/directives/n-download.directive.js
import { __decorate } from "tslib";
import { Directive, Input, Output, EventEmitter, HostListener } from '@angular/core';
import { NFileIOService } from '../services/n-fileIO.service';
var NDownloadDirective = /** @class */ (function () {
    function NDownloadDirective(fsv) {
        this.fsv = fsv;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    NDownloadDirective.prototype.methodToHandleMouseClickAction = function () {
        var _this = this;
        this.downloadFile()
            .then(function (res) { return _this.onsuccess.emit(res); })
            .catch(function (err) { return _this.onerror.emit(err); });
    };
    NDownloadDirective.prototype.downloadFile = function () {
        return this.fsv.download(this.downloadOptions);
    };
    NDownloadDirective.ctorParameters = function () { return [
        { type: NFileIOService }
    ]; };
    __decorate([
        Input()
    ], NDownloadDirective.prototype, "downloadOptions", void 0);
    __decorate([
        Output()
    ], NDownloadDirective.prototype, "onsuccess", void 0);
    __decorate([
        Output()
    ], NDownloadDirective.prototype, "onerror", void 0);
    __decorate([
        HostListener('click')
    ], NDownloadDirective.prototype, "methodToHandleMouseClickAction", null);
    NDownloadDirective = __decorate([
        Directive({ selector: '[n-download]' })
    ], NDownloadDirective);
    return NDownloadDirective;
}());
export { NDownloadDirective };
