// neutrinos-module/esm5/src/app/neutrinos-module/neutrinos-file/directives/n-scanner.directive.js
import { __decorate } from "tslib";
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { HostListener } from '@angular/core';
import { NFileIOService } from '../services/n-fileIO.service';
var NScanDirective = /** @class */ (function () {
    function NScanDirective(fsv) {
        this.fsv = fsv;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    NScanDirective.prototype.methodToHandleMouseClickAction = function () {
        var _this = this;
        this.getPicture()
            .then(function (sucess) { return _this.onsuccess.emit(sucess); })
            .catch(function (error) { return _this.onerror.emit(error); });
    };
    NScanDirective.prototype.getPicture = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.fsv.scanPicture(_this.scanOptions).then(function (res) {
                if (_this.scanOptions.doUpload) {
                    var options = {
                        'formData': res,
                        'entityName': _this.scanOptions.entityName,
                        'metadata': _this.scanOptions.metadata
                    };
                    _this.fsv.upload(options).then(function (uri) {
                        resolve(uri);
                    }).catch(function (err) { return reject(err); });
                }
                else {
                    resolve(res);
                }
            }).catch(function (err) {
                return reject(err);
            });
        });
    };
    NScanDirective.ctorParameters = function () { return [
        { type: NFileIOService }
    ]; };
    __decorate([
        Input()
    ], NScanDirective.prototype, "scanOptions", void 0);
    __decorate([
        Output()
    ], NScanDirective.prototype, "onsuccess", void 0);
    __decorate([
        Output()
    ], NScanDirective.prototype, "onerror", void 0);
    __decorate([
        HostListener('click')
    ], NScanDirective.prototype, "methodToHandleMouseClickAction", null);
    NScanDirective = __decorate([
        Directive({
            selector: '[n-scan]'
        })
    ], NScanDirective);
    return NScanDirective;
}());
export { NScanDirective };
