// neutrinos-module/esm5/src/app/neutrinos-module/neutrinos-file/directives/n-shake.directive.js
import { __decorate } from "tslib";
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { HostListener } from '@angular/core';
import { NFileIOService } from '../services/n-fileIO.service';
var NShakeDirective = /** @class */ (function () {
    function NShakeDirective(fsv) {
        this.fsv = fsv;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    NShakeDirective.prototype.methodToHandleMouseClickAction = function () {
        var _this = this;
        this.getShake()
            .then(function (sucess) { return _this.onsuccess.emit(sucess); })
            .catch(function (error) { return _this.onerror.emit(error); });
    };
    NShakeDirective.prototype.getShake = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.fsv.getShake(_this.shakeOptions).then(function (res) {
                return resolve(res);
            }).catch(function (err) { return reject(err); });
        });
    };
    NShakeDirective.ctorParameters = function () { return [
        { type: NFileIOService }
    ]; };
    __decorate([
        Input()
    ], NShakeDirective.prototype, "shakeOptions", void 0);
    __decorate([
        Output()
    ], NShakeDirective.prototype, "onsuccess", void 0);
    __decorate([
        Output()
    ], NShakeDirective.prototype, "onerror", void 0);
    __decorate([
        HostListener('click')
    ], NShakeDirective.prototype, "methodToHandleMouseClickAction", null);
    NShakeDirective = __decorate([
        Directive({
            selector: '[n-shake]'
        })
    ], NShakeDirective);
    return NShakeDirective;
}());
export { NShakeDirective };
