// neutrinos-module/esm5/src/app/neutrinos-module/neutrinos-file/directives/n-tts.directive.js
import { __decorate } from "tslib";
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { HostListener } from '@angular/core';
import { NFileIOService } from '../services/n-fileIO.service';
var NTtsDirective = /** @class */ (function () {
    function NTtsDirective(fsv) {
        this.fsv = fsv;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    NTtsDirective.prototype.methodToHandleMouseClickAction = function () {
        var _this = this;
        this.getTts()
            .then(function (sucess) { return _this.onsuccess.emit(sucess); })
            .catch(function (error) { return _this.onerror.emit(error); });
    };
    NTtsDirective.prototype.getTts = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.fsv.getTts(_this.ttsOptions).then(function (res) {
                return resolve(res);
            }).catch(function (err) { return reject(err); });
        });
    };
    NTtsDirective.ctorParameters = function () { return [
        { type: NFileIOService }
    ]; };
    __decorate([
        Input()
    ], NTtsDirective.prototype, "ttsOptions", void 0);
    __decorate([
        Output()
    ], NTtsDirective.prototype, "onsuccess", void 0);
    __decorate([
        Output()
    ], NTtsDirective.prototype, "onerror", void 0);
    __decorate([
        HostListener('click')
    ], NTtsDirective.prototype, "methodToHandleMouseClickAction", null);
    NTtsDirective = __decorate([
        Directive({
            selector: '[n-tts]'
        })
    ], NTtsDirective);
    return NTtsDirective;
}());
export { NTtsDirective };
